apps/daimo-web/src/components/PerformWalletAction.tsx

```diff
@@ -0,0 +1,175 @@
+"use client";
+import {
+  DaimoLinkStatus,
+  DaimoNoteStatus,
+  DaimoRequestStatus,
+  getNoteClaimSignature,
+} from "@daimo/common";
+import {
+  ephemeralNotesABI,
+  ephemeralNotesAddress,
+  tokenMetadata,
+} from "@daimo/contract";
+import { ConnectButton } from "@rainbow-me/rainbowkit";
+import { useEffect, useMemo, useState } from "react";
+import { Address, Hex, InsufficientFundsError, parseUnits } from "viem";
+import {
+  erc20ABI,
+  useAccount,
+  useContractWrite,
+  useNetwork,
+  usePrepareContractWrite,
+} from "wagmi";
+
+import { H2 } from "./typography";
+
+type Action = {
+  wagmiPrep: {
+    address: Address;
+    abi: readonly unknown[];
+    functionName: string;
+    args: readonly unknown[];
+  };
+  buttonTitle: string;
+};
+
+async function linkStatusToAction(
+  linkStatus: DaimoLinkStatus,
+  selfAddress: Address,
+  ephPrivateKey: Hex | undefined
+): Promise<Action> {
+  switch (linkStatus.link.type) {
+    case "request": {
+      const { recipient } = linkStatus as DaimoRequestStatus;
+      const parsedAmount = parseUnits(
+        linkStatus.link.dollars,
+        tokenMetadata.decimals
+      );
+      return {
+        wagmiPrep: {
+          address: tokenMetadata.address,
+          abi: erc20ABI as readonly unknown[],
+          functionName: "transfer" as const,
+          args: [recipient.addr, parsedAmount] as const,
+        },
+        buttonTitle: "Send via Wallet",
+      };
+    }
+    case "note": {
+      const { sender } = linkStatus as DaimoNoteStatus;
+      const signature = await getNoteClaimSignature(
+        sender.addr,
+        selfAddress,
+        ephPrivateKey
+      );
+
+      return {
+        wagmiPrep: {
+          address: ephemeralNotesAddress,
+          abi: ephemeralNotesABI as readonly unknown[],
+          functionName: "claimNote" as const,
+          args: [linkStatus.link.ephemeralOwner, signature] as const,
+        },
+        buttonTitle: "Claim via Wallet",
+      };
+    }
+    default: {
+      throw new Error(
+        `unexpected DaimoLinkStatus ${linkStatus.link.type} for wallet action`
+      );
+    }
+  }
+}
+
+export function PerformWalletAction({
+  linkStatus,
+}: {
+  linkStatus: DaimoLinkStatus;
+}) {
+  const { address, isConnected } = useAccount();
+  const { chain } = useNetwork();
+
+  const [action, setAction] = useState<Action>();
+  const [creationError, setCreationError] = useState<string>();
+  const { config, error } = usePrepareContractWrite(action?.wagmiPrep);
+  const { data, isLoading, isSuccess, write } = useContractWrite(config);
+
+  const humanReadableError = useMemo(() => {
+    if (creationError !== undefined) return creationError;
+
+    if (!error) return undefined;
+    if (error.message.match(/ERC20: transfer amount exceeds balance/)) {
+      return "Not enough USDC in wallet";
```

Comment:
this is the not-enough-$ to claim a link problem?

Reply:
requests: not enough USDC to fulfil a request